### PR TITLE
[AWS]Add AwsKmsClient for table encryption KMS client implementation

### PR DESCRIPTION
**Pull request by JiJiTang**
Aws KMS client implementation:

- Support data key generation using AWS KMS, configurable via property `ENABLE_KEY_GENERATION`

### CODE REVIEW

**Comment by ggershinsky** [comment]
cc @jackye1995 @RussellSpitzer @flyrain

File: aws/src/main/java/org/apache/iceberg/aws/AwsKmsClient.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import software.amazon.awssdk.services.kms.model.GenerateDataKeyRequest;
+import software.amazon.awssdk.services.kms.model.GenerateDataKeyResponse;
+
+public class AwsKmsClient implements KmsClient {
+  /*
+   * property for enabling data key generation using AWS KMS service, by default it's enabled,
+   * please set this property to "false" to disable key generation by AWS KMS server.
+   */
+  public static final String ENABLE_KEY_GENERATION = "kms.client.aws.key.generation.enabled";
+  // property for specifying data key generation key spec: AES_256 or AES_128
```
Comment (flyrain) [inline aws/src/main/java/org/apache/iceberg/aws/AwsKmsClient.java:41]: Nit: suggest an empty line above the comment.

**Review by flyrain — APPROVED** [review]
LGTM

**Comment by JiJiTang** [comment]
@danielcweeks , thanks VM for the good comments.  Please find my comments as the following:

1. Default AWS KMS client created by default AWSClientFactory relies on normal AWS credential provider which works out of box for typical AWS environment without the need of customoisation.
2. I didn't use `AwsClientFactories.from`  because  AWSClientFactory relies on catalog properties for initialisation but Iceberg Encryption KMS client `initialise` method only brings table properties not catalog properties in the scope.

IMHO, support key generation or not should be per table configuration and configured via table properties, not sure we should configure via AWSClientFactory, cc @ggershinsky for more inputs.

**Comment by ggershinsky** [comment]
Yep, a KMS client doesn't receive catalog properties (which makes sense, because a KMS and a Catalog are two unrelated services). It can be configured via table properties and via custom channels. If the default AWS KMS client works out of box for a typical environment, then maybe we can start without custom channels; allowing users to override this as needed.

**Comment by ggershinsky** [comment]
not any users, but org security experts, see eg our earlier docs at https://spark.apache.org/docs/latest/sql-data-sources-parquet.html#kms-client . Let us think a bit more about this PR.

**Pull request closed without merging**